sleepy/types.py

```diff
@@ -613,48 +617,35 @@ def make_extract_val(self, union_ir_val: ir.Value,
     return context.builder.load(untagged_union_ptr, name=name)
 
   def copy_with_narrowed_types(self, narrow_to_types: List[Type] | Set[Type]) -> UnionType:
-    possible_types = [possible_type for possible_type in self.possible_types if possible_type in narrow_to_types]
-    possible_type_nums = [
-      possible_type_num
-      for possible_type, possible_type_num in zip(self.possible_types, self.possible_type_nums)
-      if possible_type in narrow_to_types]
-    return UnionType(possible_types=possible_types, possible_type_nums=possible_type_nums, val_size=self.val_size)
-
-  def copy_with_extended_types(self, extended_types: List[Type],
-                               extended_type_nums: Union[List[Union[int, None]], None] = None) -> UnionType:
-    assert all(not isinstance(extended_type, UnionType) for extended_type in extended_types)
-    if extended_type_nums is None:
-      extended_type_nums = [None] * len(extended_types)
-    assert len(extended_types) == len(extended_type_nums)
-    new_possible_types = self.possible_types.copy()
-    new_possible_type_nums = self.possible_type_nums.copy()
-    for extended_type, extended_type_num in zip(extended_types, extended_type_nums):
-      if extended_type in new_possible_types:
-        continue
-      new_possible_types.append(extended_type)
-      if extended_type_num is not None and extended_type_num not in new_possible_type_nums:
-        # noinspection PyTypeChecker
-        new_possible_type_nums.append(extended_type_num)
-      else:
-        next_type_num = max(new_possible_type_nums) + 1 if len(new_possible_type_nums) > 0 else 0
-        new_possible_type_nums.append(next_type_num)
-    if self.val_size is None or any(
-      extended_type.has_unfilled_template_parameters() for extended_type in extended_types):
-      new_val_size = None
-    else:
-      new_val_size = max([self.val_size] + [extended_type.size for extended_type in extended_types])
     return UnionType(
-      possible_types=new_possible_types, possible_type_nums=new_possible_type_nums, val_size=new_val_size)
+      type_mapping=frozendict((t, self.type_mapping[t]) for t in narrow_to_types),
```

Comment:
Do we also need the case where `narrow_to_types` is not a subset of `self.type_mapping.keys()`? I think we need to support that (but interestingly no test cases seem to fail. Maybe if you have some `x: Int|Double; assert(x is Str)` or so it would fail.